### PR TITLE
Add attachments from the junit-attachments plugin into Jira attachments

### PR DESCRIPTION
**Pull request by imonteroperez**
### Highlights
- Before this PR the attachments added to the Jira tickets are always the stderr, stdout and error stacktrace and details
- After this PR, if the test result has additional attachments provided by third-party tools (like the junit-attachments publisher) these would be also added to the Jira ticket.

### Testing done
- Tested with Jira and Jenkins locally and working as expected

- Jenkins attachment

Captura de pantalla de 2024-04-21 21-30-54

- Jira attachment (contains the one from the junit-attachment in addition to the current ones - stderr, stdout, etc.)

Captura de pantalla de 2024-04-21 21-30-34

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraTestDataPublisher.java [hunk 1]
```diff
@@ -56,16 +57,21 @@
...
+import java.util.Map;
 
 /**
  * Created by tuicu.
  */
 public class JiraTestDataPublisher extends TestDataPublisher {
 
 	public static final boolean DEBUG = false;
+	
+	/** Attachments obtained from junit-attachments plugin if installed **/
```
Comment (imonteroperez) [inline src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraTestDataPublisher.java:73]:
```suggestion
	/** Attachments obtained from junit-attachments plugin indexed by className and test method name **/
```

File: src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraUtils.java [hunk 2]
```diff
@@ -230,6 +233,20 @@ private static String createIssueInput(IssueInput issueInput, CaseResult test) {
...
+            if (attachmentStorage.exists()) {
+                FilePath attachmentPath = AttachmentPublisher.getAttachmentPath(attachmentStorage, test.getClassName());
+                for (FilePath file : attachmentPath.list()) {
+                    if (attachments.contains(file.getName())) {
+                        issueClient.addAttachment(attachmentsUri, file.read(), file.getName()).claim();
+                    }
+                }
+            }
+        } catch (IOException | InterruptedException e) {
+            JiraUtils.log(String.format("Unable to access to attachment storage for %s:%s", test.getName(), test.getRun().getId()));
```
Comment (cathychan) [inline src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraUtils.java:248]:
```suggestion
            JiraUtils.log(String.format("Unable to access attachment storage for %s:%s", test.getName(), test.getRun().getId()));
```

**Review by cathychan — COMMENTED** [review]
IMO this should be an opt-in feature. If I have some builds that have massive test attachments, I may not want to include them in the Jira tickets.

**Comment by imonteroperez** [comment]
> IMO this should be an opt-in feature. If I have some builds that have massive test attachments, I may not want to include them in the Jira tickets.

LGTM, will add a configuration option

**Comment by imonteroperez** [comment]
> > IMO this should be an opt-in feature. If I have some builds that have massive test attachments, I may not want to include them in the Jira tickets.
> 
> LGTM, will add a configuration option

Added a configuration option via `EnvVar` on https://github.com/jenkinsci/JiraTestResultReporter-plugin/pull/144/commits/ce6f0c7c74142eccd9ed76e7e904f1a5044044a7. I discarded to include an option to the step to keep it simple and keep backward compatibility. Also, the step options are there to specify the behavior (raise, resolve, etc.) but here we want a specific additional operation on raising Jira issues, so that is the reason why I decided to go through an env var.

File: src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraTestDataPublisher.java [hunk 3]
```diff
@@ -56,16 +57,24 @@
...
 import java.util.List;
+import java.util.Map;
 
 /**
  * Created by tuicu.
  */
 public class JiraTestDataPublisher extends TestDataPublisher {
 
-	public static final boolean DEBUG = false;
+    public static final boolean DEBUG = false;
```
Comment (imonteroperez) [inline src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraTestDataPublisher.java:71]:
```suggestion
	public static final boolean DEBUG = false;
```

File: README.md [hunk 4]
```diff
@@ -135,6 +135,12 @@ Finally, your issues are created and you can see them by clicking the links dire
 ![](img/jira-issue.png)
 
 
+### Additional configuration
+
+You can use additional environment variables to define additional configuration needs:
+
+ * `JIRA_INCLUDE_JUNIT_ATTACHMENTS`: it will include as attachments of a Jira issue additional files from [junit-attachments](https://github.com/jenkinsci/junit-attachments-plugin) plugin. Only on raising new issues.
```
Comment (jglick) [inline README.md:142]: Why an environment variable and not just a regular `boolean` data bound setter on the publisher?

File: src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraTestDataPublisher.java [hunk 5]
```diff
@@ -199,6 +225,10 @@ public TestResultAction.Data contributeTestData(Run<?, ?> run, @Nonnull FilePath
         }
 
         if(JobConfigMapping.getInstance().getAutoRaiseIssue(project)) {
+            if (Boolean.getBoolean(System.getenv(JIRA_INCLUDE_JUNIT_ATTACHMENTS))) {
+                GetTestDataMethodObject methodObject = new GetTestDataMethodObject(run, workspace, launcher, listener, testResult);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraTestDataPublisher.java:230]: TIL: https://github.com/jenkinsci/junit-attachments-plugin/blob/7ca5b7a98b1c215fcd2aa21a1510c03f1b488466/src/main/java/hudson/plugins/junitattachments/GetTestDataMethodObject.java#L31-L32 does not seem to accurately describe what that class does at all.

File: src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraTestDataPublisher.java [hunk 6]
```diff
@@ -178,7 +204,7 @@ public TestResultAction.Data contributeTestData(Run<?, ?> run, @Nonnull FilePath
                                                     throws IOException, InterruptedException {
         
 	EnvVars envVars = run.getEnvironment(listener);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraTestDataPublisher.java:206]: This gets the top-level environment for the build. `withEnv` does not affect it.

**Review by jglick — CHANGES_REQUESTED** [review]

File: src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraTestDataPublisher.java (hunk 6 above)
Comment (imonteroperez) [reply src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraTestDataPublisher.java:206]: You are right, actually, I'm testing this right now and detected this issue. TBH rethinking on to setup a boolean data bound setter better than this.

File: README.md (hunk 4 above)
Comment (imonteroperez) [reply README.md:142]: At the very beginning considered this https://github.com/jenkinsci/JiraTestResultReporter-plugin/pull/144#issuecomment-2071625826 but I'm reconsidering this option and return to the boolean data bound setter on the publisher

File: src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraTestDataPublisher.java (hunk 5 above)
Comment (imonteroperez) [reply src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraTestDataPublisher.java:230]: It retrieves all the attachments per test class and method

File: README.md (hunk 4 above)
Comment (jglick) [reply README.md:142]: All configuration of Pipeline steps, or structs used inside steps (this case), should be via regular data binding. It should be self-documenting as well, using `config.jelly`. See [docs](https://www.jenkins.io/doc/developer/plugin-development/pipeline-integration/).

File: src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraTestDataPublisher.java (hunk 6 above)
Comment (imonteroperez) [reply src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraTestDataPublisher.java:205]: Addressed in https://github.com/jenkinsci/JiraTestResultReporter-plugin/pull/144/commits/a384e3c2c877576763c7137ef14488508927a1d9

File: README.md (hunk 4 above)
Comment (imonteroperez) [reply README.md]: Done in https://github.com/jenkinsci/JiraTestResultReporter-plugin/pull/144/commits/a384e3c2c877576763c7137ef14488508927a1d9

File: src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraTestDataPublisher.java [hunk 7]
```diff
@@ -130,7 +153,8 @@ private JobConfigMapping.JobConfigEntry getJobConfig() {
      */
 	@DataBoundConstructor
 	public JiraTestDataPublisher(List<AbstractFields> configs, String projectKey, String issueType,
-                                 boolean autoRaiseIssue, boolean autoResolveIssue, boolean autoUnlinkIssue, boolean overrideResolvedIssues) {
+                                 boolean autoRaiseIssue, boolean autoResolveIssue, boolean autoUnlinkIssue, boolean overrideResolvedIssues, 
+                                 boolean additionalAttachments) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraTestDataPublisher.java:157]: Better to use `@DataBoundSetter`. https://www.jenkins.io/doc/developer/plugin-development/pipeline-integration/#constructor-vs-setters

Comment (imonteroperez) [reply src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraTestDataPublisher.java:157]: I will do this refactoring outside the scope of this PR. Not sure which parameters can be delegated to a setter given documentation said this is for optional ones.

Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraTestDataPublisher.java:157]: Well the _new_ option you just added is clearly optional and so it should be added via a `@DataBoundSetter`. Refactoring _existing_ options is another matter, but you should never be adding parameters to an existing `@DataBoundConstructor`.

Comment (imonteroperez) [reply src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraTestDataPublisher.java]: Addresed for the new optional field in https://github.com/jenkinsci/JiraTestResultReporter-plugin/pull/144/commits/d96ed87ceeefc981a585342cae000b53a2b0800d

File: src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraTestDataPublisher.java [hunk 8]
```diff
@@ -121,12 +145,32 @@ private boolean isPipelineInvocation() {
     private JobConfigMapping.JobConfigEntry getJobConfig() {
         return jobConfig;
     }
+    
+    @DataBoundSetter
+    public void setAdditionalAttachments(boolean additionalAttachments) {
```
[the remark this answers is not in the document]
Comment (imonteroperez) [reply src/main/java/org/jenkinsci/plugins/JiraTestResultReporter/JiraTestDataPublisher.java:157]: Addressed on https://github.com/jenkinsci/JiraTestResultReporter-plugin/pull/144/commits/86b36798e27cbd6653295a21c0c02cae5b76d25f

**Review by jglick — APPROVED** [review]
Looks OK.

**Pull request merged**